modules/samples/firealarm/component/plugin/src/main/java/org/homeautomation/firealarm/plugin/impl/dao/DeviceTypeDAO.java

```diff
@@ -0,0 +1,132 @@
+            Context ctx = new InitialContext();
+            dataSource = (DataSource) ctx.lookup(DeviceTypeConstants.DATA_SOURCE_NAME);
+        } catch (NamingException e) {
+            log.error("Error while looking up the data source: " +
+                      DeviceTypeConstants.DATA_SOURCE_NAME);
+        }
+
+    }
+
+    public static void beginTransaction() throws FirealarmPluginException {
+        try {
+            Connection conn = dataSource.getConnection();
+            conn.setAutoCommit(false);
+            currentConnection.set(conn);
+        } catch (SQLException e) {
+            throw new FirealarmPluginException(
+                    "Error occurred while retrieving datasource connection", e);
+        }
+    }
+
+    public static Connection getConnection() throws FirealarmPluginException {
+        if (currentConnection.get() == null) {
+            try {
+                currentConnection.set(dataSource.getConnection());
+            } catch (SQLException e) {
+                throw new FirealarmPluginException(
+                        "Error occurred while retrieving data source connection", e);
+            }
+        }
+        return currentConnection.get();
+    }
+
+    public static void commitTransaction() throws FirealarmPluginException {
+        try {
+            Connection conn = currentConnection.get();
+            if (conn != null) {
+                conn.commit();
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Datasource connection associated with the current thread is null, " +
+                              "hence commit has not been attempted");
+                }
+            }
+        } catch (SQLException e) {
+            throw new FirealarmPluginException(
+                    "Error occurred while committing the transaction", e);
+        } finally {
+            closeConnection();
+        }
+    }
+
+    public static void closeConnection() throws FirealarmPluginException {
+
+        Connection con = currentConnection.get();
+        if (con != null) {
+            try {
+                con.close();
+            } catch (SQLException e) {
+                log.error("Error occurred while close the connection");
+            }
+        }
+        currentConnection.remove();
+    }
+
+    public static void rollbackTransaction() throws FirealarmPluginException {
+        try {
+            Connection conn = currentConnection.get();
+            if (conn != null) {
+                conn.rollback();
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(
+                            "Datasource connection associated with the current thread is null, " +
+                            "hence rollback has not been attempted");
+                }
+            }
+        } catch (SQLException e) {
+            throw new FirealarmPluginException("Error occurred while rollback the transaction", e);
```

Comment:
How can this DeviceTypeDAO return a FirealarmPluginException?

And ideally, my understanding is that these core DAO functions should be placed within a DAO factory class.